include/deal.II/lac/scalapack.h

```diff
@@ -0,0 +1,478 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2017 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_scalapack_h
+#define dealii_scalapack_h
+
+#include <deal.II/base/config.h>
+
+#ifdef DEAL_II_WITH_SCALAPACK
+
+#include <deal.II/base/exceptions.h>
+#include <deal.II/lac/full_matrix.h>
+#include <deal.II/lac/lapack_support.h>
+#include <deal.II/base/mpi.h>
+#include <mpi.h>
+
+#include <memory>
+
+DEAL_II_NAMESPACE_OPEN
+
+
+/**
+ * Forward declaration of class ScaLAPACKMatrix
+ */
+template <typename NumberType> class ScaLAPACKMatrix;
+
+
+/**
+ * A class taking care of setting up the process grid for BLACS
+ * ScaLapack matrices will have shared pointer of this class to perform block-cyclic distribution
+ *
+ * @author Benjamin Brands, 2017
+ */
+class ProcessGrid
+{
+public:
+
+  /**
+   * Declare class ScaLAPACK as friend to provide access to private members, e.g. the MPI Communicator
+   */
+  template <typename NumberType> friend class ScaLAPACKMatrix;
+
+  /*
+   * Constructor for a process grid for a given @p mpi_communicator
+   * The pair @p grid_dimensions contains the user-defined numbers of process rows and columns
+   */
+  ProcessGrid(MPI_Comm mpi_communicator, const std::pair<int,int> &grid_dimensions);
+
+  /*
+   * Constructor for a process grid for a given @p mpi_communicator
+   * The process grid is built based on the dimension and blockcyclic distribution of a target matrix described by @p matrix_dimensions and @p block_sizes
+   */
+  ProcessGrid(MPI_Comm mpi_communicator,
+              const std::pair<int,int> &matrix_dimensions, const std::pair<int,int> &block_sizes);
+
+  /**
+  * Destructor
+  */
+  virtual ~ProcessGrid();
+
+  /**
+  * Return the number of rows in processes grid.
+  */
+  int get_process_grid_rows() const;
```

Comment:
Can these be negative? If not use ``unsigned int`` (and all other places this is exposed)

Reply:
it can not, but ScaLAPACK functions don't know how to work with `unsigned int` so we sticked with `int` throughout.

Reply:
> You could just cast these to int as necessary

I don't think i can in case of pointers.